File: packages/epidemica_proximity_ios/wire/Sources/ProximityWire/ProximityPayload.swift
import Foundation

/// The bytes two Epidemica devices exchange over Bluetooth.
///
/// Specified in `contracts/wire/proximity_payload/1.0.0.md` and held to the shared vectors in
/// `1.0.0.vectors.json`, which the Kotlin implementation decodes too. Two implementations that can
/// only ever meet on real hardware are otherwise free to disagree about byte order or trailing
/// data, and the symptom would be an iPhone and an Android phone sitting side by side, detecting
/// nothing, with no error anywhere.
public enum ProximityPayload {

    public static let formatVersion: UInt8 = 1
    public static let sizeBytes = 18

    private static let offsetVersion = 0
    private static let offsetPseudonym = 1
    private static let offsetDeviceClass = 17

    private static let deviceClassIOS: UInt8 = 0
    private static let deviceClassAndroid: UInt8 = 1
    private static let deviceClassUnknown: UInt8 = 255

    public struct Decoded: Equatable {
        public let pseudonym: String
        public let deviceClass: String?

        public init(pseudonym: String, deviceClass: String?) {
            self.pseudonym = pseudonym
            self.deviceClass = deviceClass
        }
    }

    public static func encode(pseudonym: String, deviceClass: String?) -> Data? {
        guard let identity = uuidToBytes(pseudonym) else { return nil }
        var out = Data(capacity: sizeBytes)
        out.append(formatVersion)
        out.append(contentsOf: identity)
        switch deviceClass {
        case "ios": out.append(deviceClassIOS)
        case "android": out.append(deviceClassAndroid)
        default: out.append(deviceClassUnknown)
        }
        return out
    }

    /// Nil when the payload is not one this build can read.
    public static func decode(_ data: Data?) -> Decoded? {
        // Trailing bytes are tolerated: a later version may append fields, and rejecting on
        // unexpected length would make this build blind to every device running that version.
        guard let data, data.count >= sizeBytes else { return nil }
        let bytes = [UInt8](data)
        guard bytes[offsetVersion] == formatVersion else { return nil }

        let pseudonym = uuidFromBytes(Array(bytes[offsetPseudonym..<(offsetPseudonym + 16)]))
        let deviceClass: String?
        switch bytes[offsetDeviceClass] {
        case deviceClassIOS: deviceClass = "ios"
        case deviceClassAndroid: deviceClass = "android"
        default: deviceClass = nil
        }
        return Decoded(pseudonym: pseudonym, deviceClass: deviceClass)
    }

    private static func uuidToBytes(_ uuid: String) -> [UInt8]? {
        guard let parsed = UUID(uuidString: uuid) else { return nil }
        let u = parsed.uuid
        return [
            u.0, u.1, u.2, u.3, u.4, u.5, u.6, u.7,
            u.8, u.9, u.10, u.11, u.12, u.13, u.14, u.15,
        ]
    }

    private static func uuidFromBytes(_ bytes: [UInt8]) -> String {
        let hex = bytes.map { String(format: "%02x", $0) }.joined()
        let part = { (lower: Int, upper: Int) -> String in
            let start = hex.index(hex.startIndex, offsetBy: lower)
            let end = hex.index(hex.startIndex, offsetBy: upper)
            return String(hex[start..<end])
        }
        return "\(part(0, 8))-\(part(8, 12))-\(part(12, 16))-\(part(16, 20))-\(part(20, 32))"
    }
}

File: packages/epidemica_proximity_android/android/src/main/kotlin/info/epidemica/proximity/ProximityPayload.kt
package info.epidemica.proximity

import java.util.Locale
import java.util.UUID

/**
 * The bytes two Epidemica devices exchange over Bluetooth.
 *
 * Specified in `contracts/wire/proximity_payload/1.0.0.md` and held to the shared vectors in
 * `1.0.0.vectors.json`, which the Swift implementation decodes too. Pure: no Herald types, no
 * Android types, so it runs as a plain JVM unit test.
 */
object ProximityPayload {

    const val FORMAT_VERSION = 1
    const val SIZE_BYTES = 18

    private const val OFFSET_VERSION = 0
    private const val OFFSET_PSEUDONYM = 1
    private const val OFFSET_DEVICE_CLASS = 17

    private const val DEVICE_CLASS_IOS = 0
    private const val DEVICE_CLASS_ANDROID = 1
    private const val DEVICE_CLASS_UNKNOWN = 255

    data class Decoded(val pseudonym: String, val deviceClass: String?)

    fun encode(pseudonym: String, deviceClass: String?): ByteArray {
        val out = ByteArray(SIZE_BYTES)
        out[OFFSET_VERSION] = FORMAT_VERSION.toByte()
        uuidToBytes(pseudonym).copyInto(out, OFFSET_PSEUDONYM)
        out[OFFSET_DEVICE_CLASS] = when (deviceClass) {
            "ios" -> DEVICE_CLASS_IOS
            "android" -> DEVICE_CLASS_ANDROID
            else -> DEVICE_CLASS_UNKNOWN
        }.toByte()
        return out
    }

    /** Null when the payload is not one this build can read. */
    fun decode(bytes: ByteArray?): Decoded? {
        // Trailing bytes are tolerated: a later version may append fields, and rejecting on
        // unexpected length would make this build blind to every device running that version.
        if (bytes == null || bytes.size < SIZE_BYTES) return null
        if (bytes[OFFSET_VERSION].toInt() and 0xff != FORMAT_VERSION) return null

        val pseudonym = uuidFromBytes(bytes, OFFSET_PSEUDONYM)
        val deviceClass = when (bytes[OFFSET_DEVICE_CLASS].toInt() and 0xff) {
            DEVICE_CLASS_IOS -> "ios"
            DEVICE_CLASS_ANDROID -> "android"
            else -> null
        }
        return Decoded(pseudonym, deviceClass)
    }

    private fun uuidToBytes(uuid: String): ByteArray {
        val parsed = UUID.fromString(uuid)
        val out = ByteArray(16)
        var high = parsed.mostSignificantBits
        var low = parsed.leastSignificantBits
        for (i in 7 downTo 0) {
            out[i] = (high and 0xff).toByte()
            high = high ushr 8
            out[i + 8] = (low and 0xff).toByte()
            low = low ushr 8
        }
        return out
    }

    private fun uuidFromBytes(bytes: ByteArray, offset: Int): String {
        val hex = StringBuilder(32)
        for (i in 0 until 16) {
            hex.append(String.format(Locale.ROOT, "%02x", bytes[offset + i].toInt() and 0xff))
        }
        return "${hex.substring(0, 8)}-${hex.substring(8, 12)}-${hex.substring(12, 16)}-" +
            "${hex.substring(16, 20)}-${hex.substring(20, 32)}"
    }
}
